---
title: Analysis of variable DNA loops with diffloop
author: Caleb Lareau & Martin Aryee
date: "`r Sys.Date()`"
output:
  rmarkdown::html_document:
    theme: flatly
vignette: >
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---
***

# Overview
The purpose of this document is to describe in greater detail much of the useful functionality implemented in `diffloop`, an `R/Bioconductor` package. As the name suggests, the goal of this package was to create a computational framework to identify differential loops between conditions akin to differential gene expression analyses from RNA-Seq and related transcriptomics data. However, much of the more useful methodology is incorporated *before* one runs differential association (_i.e._ quality control) and after differential loops are called (_e.g._ visualization, correlating differential loops with epigenomic signal and gene expression, etc.). This extended online resource provides a means for understanding these function

Additional steps regarding the quality control ana analysis of similar ENCODE ChIA-PET data can be found [here](https://rpubs.com/caleblareau/diffloop_vignette). 

## On preprocessing

Though outside the scope of `diffloop`, an important component to consider before using this package is the choice of preprocessing associated with the loop data. For this vignette, we used the [mango](https://github.com/dphansti/mango) ChIA-PET pre-processing pipeline importantly specifying the `reportAllPairs = TRUE` flag such that any PET binding two peaks would be retained. However, it may be optimal to process the interaction set together, which requires pre-specifying anchor loci. To this end, we recommend an in-house tool, [hichipper](https://aryeelab.org/hichipper), which enables this and other HiChIP-specific methods.

Regardless of the choice of preprocessing, some basic assumptions about the anchor size have to be made. In most Hi-C pre-processing pipelines, a fixed genome window of 5kb, 20kb, or similar distance is employed. From what we've seen using preprocessing tools, most support some fixed-width padding around a `.narrowPeak` file that is called from `macs2` or a similar peak-caller when an immunoprecipitation is used (in HiChIP and ChIA-PET). Tools such as `mango` and `hichipper` use 500-1,000 bp as anchor pads, though users should consider the question of interest before selecting window padding, etc., for their experimental system. While `diffloop` has functionality to enable anchor padding, in general, this is best employed early in the pre-processing steps using the flags made available in these and other tools. 

## On single-replicate samples 

Akin to `DESeq`, `edgeR`, and related RNA analysis frameworks, replicate samples are essentially *required* to accurately identify differential loops with statistical confidence. However, one can certainly utilize `diffloop` to identify putative differential loops in single replicate settings. For example, suppose you had a `loops()` object called `dat` with 2 samples that that you wanted to compare against each other. While the assumptions needed for variance estimation in the loop counts make computing a p-value or FDR per-loop impossible as we will demonstrate later with replicates, we'd advice the following workflow for find differential loops in single-replicate settings.

```{r singles, eval=FALSE, message=FALSE, warning = FALSE}
dat <- loopsMake("../loopsdata")
dat@rowData$diffCount <- dat@counts[,2] - dat@counts[,1] # per loop differences between sample 1 and 2
df <- summary(dat)
```

**Note:** above are real diffloop commands but are not executable in the lense of the vignette. 

In the code above, the idea is to add a column to the `rowData` slot that is the difference in counts per loop between sample 2 and sample 2, which may be informative for finding differential loops between the two samples. the `summary` command in `diffloop` (when applied to a `loops()` object) results in a `data.frame` of useful information per loop where users can now sort `df` by the `diffCount` column much like any other `data.frame` in R. Again, while single replicate conditions are certainly sub-optimal, we suggest this framework for a simple first pass in this setting. 

## Sample analysis of ENCODE ChIA-PET data

The raw `.fastq` read files for the POL2 data were preprocessed with the [mango](https://github.com/dphansti/mango) package, an R package that relies on bowtie, bedtools, and MACS2 to align reads, call anchor peaks, and summarize PETs per sample in the ChIA-PET experiment.

```
head -3 *.interactions.all.mango
```
yields-- 
```
chr1    32705184    32708527    chr1    32712484    32715408    4   1
chr1    32705184    32708527    chr1    32755920    32759477    3   1
chr1    32705184    32708527    chr1    32799616    32803060    6   1
```

Another possible format for reads are `.bedpe` files, which are very similar to the output files produced by mango. To use the loopsMake function from a different preprocessing step, have files X.loop_counts.bedpe,Y.loop_counts.bedpe, Z.loop_counts.bedpe in bed_dir for samples = (X,Y,Z) where the first lines should resemble:

```
head -3 *.loop_counts.bedpe
```
yields--
```
1 10002272 10004045 10 120968807 120969483 . 1
1 10002272 10004045 10 99551498 99552470 . 1
1 10002272 10004045 1 10002272 10004045 . 17
```

where the first three columns specify the position (chr:start:stop) of the first anchor, the second three columns specify the position (chr:start:stop) of the secnd anchor, the 7th column is “.” and the 8th column is the number of paired-end reads support that particular PET.

For this vignette, we use the `.mango` output files. 


## Packages
First, we'll load all the necessary packages for this vignette overview--

```{r packages, cache=FALSE, message=FALSE, warning = FALSE}
library(diffloop)
library(GenomicRanges)
library(plotly)
```

QC steps--

```{r importQC, cache = FALSE, message=FALSE, warning = FALSE}
full <- loopsMake.mango("../data/raw_mango")
dim(full)
full <- updateLDGroups(full, c("K562", "K562", "MCF7", "MCF7"))
samples <- c("HCT116", "K562_r1", "K562_r2", "MCF7_r1", "MCF_r2")
long <- subsetLoops(full, full@rowData$loopWidth >= 5000) # remove and loops that merged together from import
dim(long)
# Remove regions of CNV
k562.cnv <- makeGRangesFromDataFrame(
  setNames(read.table("../data/cnv/K562-CNV.bedLogR")[,1:4],
                                              c("chr", "start", "end", "type")), keep.extra.columns = TRUE)
mcf7.cnv <- makeGRangesFromDataFrame(setNames(read.table("../data/cnv/MCF7-CNV.bedLogR")[,1:4],
                                              c("chr", "start", "end", "type")), keep.extra.columns = TRUE)

cnv.regions <- union(mcf7.cnv[mcols(mcf7.cnv)$type != "normal"],
                     k562.cnv[mcols(k562.cnv)$type != "normal"])
noCNV <- removeRegion(long, rmchr(cnv.regions))
dim(noCNV)

# Filter Mango interactions
mangoSig <- mangoCorrection(noCNV)
dim(mangoSig)

# save RDS
qcLoops <- filterLoops(mangoSig, width = 5000, nreplicates = 2, nsamples = 2)
```

## Integration with transcriptome data


```{r deseq, warning = FALSE, message = FALSE}
library(DESeq2)

files <- grep("counts", list.files("../data/rna-seq/"), value = TRUE)
condition <- c("k562", "k562", "k562", "mcf7", "mcf7", "mcf7")
names <- c("k1", "k2", "k3", "m1", "m2", "m3")
sampleTable <- data.frame(sampleName = names, fileName = files, condition = condition)
dds <- DESeqDataSetFromHTSeqCount(sampleTable = sampleTable, directory = "../data/rna-seq/", design = ~condition)

dds <- dds[rowSums(counts(dds)) > 1, ]  #remove zero counts
dds$condition <- factor(dds$condition, levels = c("k562", "mcf7"))
dds <- DESeq(dds)
res <- results(dds)
res <- res[complete.cases(res), ]
resDESeq2 <- res[order(res$pvalue), ]
```

# Summary

This vignette was written to demonstrate a more detailed workflow for identifying and analyzing variabile DNA loops with `diffloop`. Moreover, as handling DNA looping data is still rather new, some of the nuances associated with preprocessing discussed above should be carefullly considered. Thought the quality control measures for a set of loops require some thought, a wide variety of options are available in `diffloop` for these to be filterd before differential testing. Ultimately, the integration of extensive epigenetic and transcriptional data to identify and annotate regions of differential loops are the core functionality of the package. The visualizations provide both large-scale trends (e.g. correlations between logFC of loops and the transcription FC) and anecdotal regions of variable topology, as shown in the plots in this vignette. As the availability of DNA looping data continues to grow and experiments are designed to test unique aspects of variable DNA interactions, we expect to add more features to diffloop to further facilitate detailed analyses


## Session info

```{r sessInfo}
sessionInfo()
```

## Citations

$^1$Hnisz, Denes, et al. "Activation of proto-oncogenes by disruption of
chromosome neighborhoods." Science (2016). <br><br>
$^2$Ji, Xiong, et al. "3D Chromosome Regulatory Landscape of Human Pluripotent
Cells." Cell stem cell (2015).<br><br>
$^3$Phanstiel, D, et al. "Mango: A bias correcting ChIA-PET analysis pipeline."
Bioinformatics (2015). <br><br>
$^4$Robinson, M, et al. "edgeR: a Bioconductor package for differential expression
analysis of digital gene expression data." Bioinformatics (2010). <br><br>
$^5$Law, C, et al. "Voom: precision weights unlock linear model analysis tools
for RNA-seq read counts." Genome Biology (2014). 